pyigm/clustering/xcorr_utils.py

```diff
@@ -322,7 +386,175 @@ def spline_sensitivity(galreal, Nmin=20, magmin=17., magmax=26., delta_mag=0.5,
     return magbins, VALS, SPL
 
 
-def random_gal(galreal, Nrand, magbins, SPL):
+def random_abs_zmnx(absreal, Nrand, zmnx, wrest, dv_Galactic=100.):
+    """From a real absorber catalog it creates a random catalog.
+    Absorbers are simply placed randomly between redshift limits
+    zmnx[0] and zmnx[1]
+
+    Parameters
+    ----------
+    absreal
+    Nrand
+    zmnx
+
+    Returns
+    -------
+
+    """
+
+    absrand = absreal.repeat(Nrand)
+
+    randz = np.random.uniform(low=zmnx[0], high=zmnx[1], size=2*len(absrand))  # Buffer to reject Galaxy
+
+    # Avoid Galactic
+    galactic = mask_galactic()
+    Galz = galactic/wrest - 1.
+    z_Gal = np.outer(np.ones_like(randz), Galz)
+    # Diff
+    diff = z_Gal - np.outer(randz, np.ones_like(Galz))
+    mdiff = np.amin(np.abs(diff), axis=1)
+    # Good ones
+    gdz = mdiff > (dv_Galactic/Ckms) / (randz+1)
+    gdi = np.where(gdz)[0]
+    absrand.ZABS = randz[gdi[0:len(absrand)]]
+    # Return
+    return absrand
+
+
+def random_abs_W(absreal, Nrand, wa, fl, er, sl=3., R=20000, FWHM=10., ion='HI'):
+    """From a real absorber catalog it creates a random catalog.  For
+    a given real absorber with (z_obs,logN_obs,b_obs) it places it at
+    a new z_rand, defined by where the line could have been
+    observed.
+
+    Input parameters:
+    ---
+    absreal: numpy rec array with the absorber catalog.
+    Nrand:   number of random lines per real one generated (integer).
+    wa:      numpy array of wavelength covered by the spectrum.
+    fl:      numpy array of normalized flux.
+    er:      numpy array of error in the normalized flux of the spectrum for
+             a given wavelength.
+    sl:      significance level for the detection of the absorption line.
+    R:       resolution of the spectrograph, assumed constant
+    FWHM:    Full-width at half maximum in pixels (assumed constant). This
+             parameter defines the smoothing scale for Wmin.
+    ion:     Name of the ion. Function only valid for HI so far.
+
+    From the error we calculate the Wmin = sl * wa * er / (1+z) / R,
+    where z = wa/w0 - 1 (w0 is the rest frame wavelenght of the
+    transition) and R is the resolution of the spectrograph. We then
+    smooth Wmin with a boxcar along FWHM pixels.
+
+    For the given absorber we transform (logN_obs,b_obs) to a W_obs assuming
+    linear part of the curve-of-growth.
+
+    We then compute the redshifts where W_obs could have been observed
+    according to the given Wmin, and place Nrand new absorbers with
+    the same properties as the given one accordingly.
```

Comment:
Regions of Galactic absorption are also avoided for the randoms.

Reply:
added